cmd/validators/main.go

```diff
@@ -120,12 +127,159 @@ type pubKeyInfo struct {
+		if err != nil {
+			return fmt.Errorf("cannot serialize transaction: %s", err)
+		}
+		out.Write(raw)
+		return nil
+	}
+}
+
+func cmdMultisigSign(in io.Reader, out io.Writer) func() error {
+	return func() error {
+		flag.Usage = func() {
+			fmt.Fprint(flag.CommandLine.Output(), `
+Sign given transaction. This is decoding a transaction data from standard
+input, adds a signature and writes back to standard output signed transaction
+content.
+
+`)
+			flag.PrintDefaults()
+		}
+		var (
+			tmAddrFl = flag.String("tm", "https://bns.NETWORK.iov.one:443", "Tendermint node address. Use proper NETWORK name.")
+			keyFl    = flag.String("key", "", "Hex encoded, private key that transaction should be signed with.")
+		)
+		flag.Parse()
+
+		if *keyFl == "" {
+			return errors.New("private key is required")
+		}
+		key, err := decodePrivateKey(*keyFl)
+		if err != nil {
+			return fmt.Errorf("cannot decode private key: %s", err)
+		}
+
+		raw, err := ioutil.ReadAll(in)
+		if err != nil {
+			return fmt.Errorf("cannot read transaction from stdin: %s", err)
+		}
+		if len(raw) == 0 {
+			return errors.New("no stdin data")
+		}
+		var tx app.Tx
+		if err := tx.Unmarshal(raw); err != nil {
+			return fmt.Errorf("cannot deserialize transaction: %s", err)
+		}
+
+		genesis, err := fetchGenesis(*tmAddrFl)
+		if err != nil {
+			return fmt.Errorf("cannot fetch genesis: %s", err)
+		}
+
+		bnsClient := client.NewClient(client.NewHTTPConnection(*tmAddrFl))
+		aNonce := client.NewNonce(bnsClient, key.PublicKey().Address())
+		if seq, err := aNonce.Next(); err != nil {
+			return fmt.Errorf("cannot get the next sequence number: %s", err)
+		} else {
+			client.SignTx(&tx, key, genesis.ChainID, seq)
+		}
+
+		if raw, err := tx.Marshal(); err != nil {
+			return fmt.Errorf("cannot serialize transaction: %s", err)
+		} else {
+			out.Write(raw)
+		}
+		return nil
+	}
+}
+
+func cmdMultisigSubmit(in io.Reader) func() error {
+	return func() error {
+		flag.Usage = func() {
+			fmt.Fprint(flag.CommandLine.Output(), `
+Read binary serialized validator change transaction from standard input and
+submit it.
+
+This command is intended to be used to submit multisig authenticated requests.
+Make sure to collect enough signatures before submitting.
+
+`)
+			flag.PrintDefaults()
+		}
+		var (
+			tmAddrFl = flag.String("tm", "https://bns.NETWORK.iov.one:443", "Tendermint node address. Use proper NETWORK name.")
+		)
+		flag.Parse()
+
+		raw, err := ioutil.ReadAll(in)
+		if err != nil {
+			return fmt.Errorf("cannot read transaction from stdin: %s", err)
+		}
+		if len(raw) == 0 {
+			return errors.New("no stdin data")
+		}
+		var tx app.Tx
+		if err := tx.Unmarshal(raw); err != nil {
+			return fmt.Errorf("cannot deserialize transaction: %s", err)
+		}
+		bnsClient := client.NewClient(client.NewHTTPConnection(*tmAddrFl))
+
+		if err := bnsClient.BroadcastTx(&tx).IsError(); err != nil {
+			return fmt.Errorf("cannot broadcast transaction: %s", err)
```

Comment:
pp: no need for the `if`, just return the err

Reply:
I need the if because I am wrapping the error. :thinking:

Reply:
Maybe this can be done easier with the new error package

Reply:
I am wondering if I cannot see something obvious here. Let me explain what I think and tell me if it makes sense.

You are both talking about below code:
```go
if err := bnsClient.BroadcastTx(&tx).IsError(); err != nil {
    return fmt.Errorf("cannot broadcast transaction: %s", err)
}
```
The above could be replaced with a shorter version:
```go
return bnsClient.BroadcastTx(&tx).IsError()
```

The second example is correct, but the returned error does not provide the same information as the in the first example. I think wrapping an error with a piece of additional information is helpful.

If the new `errors` package is merged, `fmt.Errorf` could be replaced with
```go
if err := bnsClient.BroadcastTx(&tx).IsError(); err != nil {
    return errors.Wrap(err, "cannot broadcast transaction")
}
```
But the `if` must stay. Otherwise a new error would be created, even if `IsError` call returns `nil`.

Does this make sense?

Reply:
@husio Yes, with the current API. But in the refactor, you removed a nice feature of the "legacy" errors API. Look at https://github.com/iov-one/weave/pull/308#discussion_r255426591

```golang
// Wrap safely takes any error and promotes it to a TMError.
// Doing nothing on nil or an incoming TMError.
func Wrap(err error) TMError {
	// nil or TMError are no-ops
	if err == nil {
		return nil
	}
```

Reply:
You do not need to maintain "Do nothing on TmError", but at least "return nil on nil" is a nice feature

Reply:
I understand now. I was not aware of this functionality and I also did not carry it to the new `Wrap`. I is an easy change to add it back.

_tl;dr;_ I am afraid that adding support for `Wrap(<nil>, "message")` may cause subtle bugs. But I can apply this change if this is the pattern we want to follow in `weave`.

I understand that supporting `nil` error in `Wrap` function as proposed may look appealing. It allows saving on an `if` statement when last operation done in a function can return an error. It is easier to explain with an example:

```go
func foo() error {
   if err := first(); err != nil {
       return errors.Wrap(err, "first")
   }
   return errors.Wrap(second(), "second")
}
```
When calling `second`, I can wrap the result without checking if the error is `nil`. When calling `first`, I have to test for an error anyway, because the result affects the flow.

Consider the verbose version of the very same functionality:
```go
func foo() error {
   if err := first(); err != nil {
       return errors.Wrap(err, "first")
   }
   if err := second(); err != nil {
       return errors.Wrap(err, "second")
   }
   return nil
}
```
For me it is clear what the flow is. While `errors.Wrap(second(), "second")` is very dense and expands to several operations, second example stays very simple.

I would not be surprised if most programmers asked to explain the above two examples were often wrong with the first one and always right with the second one.

Maybe it all boils down to a personal preference of clarity over expressiveness. I always tend to pick the verbose approach thinking that future Piotr will appriciate me not trying to be too smart :wink:

Reply:
It is not about the first or the second usage everywhere.
It is about the ability to choose.

I think there are cases when the second usage is cleaner. When there is a long series of checks, as you mentioned. 

I think there are also cases when the first usage is cleaner. When there is only one error return, and the overhead of needing the if statement just to wrap it will cause programmer to NOT wrap it out of laziness.

I want to make it easy for those cases, and allow the developers using the errors package to choose. I don't think there are many cases where I relied on the `nil -> nil` functionality of the legacy `Wrap()`, but definitely a few places

Reply:
Also, another example to your above...

Common (unwrapped) idiom - I think putting the logic in an if statement makes it harder to read as well.

```golang
func foo() error {
   err := first()
   if err != nil {
       return err
   }
   err = second()
   return errors.Wrap(err, "second")
```

And a wrapped version

```golang
func foo() error {
   err := first()
   if err != nil {
      return errors.Wrap(err, "first")
   }
   err = second()
   return errors.Wrap(err, "second")
}
```

Reply:
I have pushed the changes required for this https://github.com/iov-one/weave/pull/275/commits/9d1614508ab365f5638be151f40eccc58fa37198
Once new errors is merged, I can update this line.